### PR TITLE
Postgres Database Instrumentation Support

### PR DESCRIPTION
**Pull request by bobuva**
### NOTE: Intended to be Pre-Released to verify with users after 1.13.0 is released

Fixes loss of `postgres` instrumentation.

Changes proposed in this pull request:
- Adds support for the .NET Postgres library named `Npgsql`.
- Instrumented methods are:
   1. `NpgsqlCommand.ExecuteReader` (two overloads)
   2. `NpgsqlCommand.ExecuteReaderAsync` (two overloads)
   3. `NpgsqlCommand.ExecuteNonQuery`
   4. `NpgsqlCommand.ExecuteNonQueryAsync`
   5. `NpgsqlCommand.ExecuteScalar`
   6. `NpgsqlCommand.ExecuteScalarAsync`

#### TO-DO
 - [ ] Test integration for memory and performance in APM Integrations Reliability Environment

#### Notes

*Note* about an unsupported overload of `NpgsqlCommand.ExecuteReaderAsync`:
A third overload that takes three arguments and returns a `ValueTask<Npgsql.NpgsqlDataReader>` is not being instrumented because to use a `ValueTask<T>` would require support for [`.NET Standard 2.1`](https://github.com/dotnet/standard/milestone/3). This version of .NET Standard is supported in .NET Core 3.0 and later but uses a different runtime and would require, IMO, a fair amount of plumbing to get it working in our complex build and test environment. That said, we definitely want to migrate to `netstandard2.1` at some point in the future.

*Testing Note*
Some of the users reported not getting `postgres` traces when they used `Dapper`, a lightweight ORM. Dapper accesses the underlying database provider library (e.g., `Npgsql`) in a way that is not captured by our `IDbCommand` and `DbCommand` integrations. By supporting the `Npgsql` Postgres library directly, we provide support for applications that use `Dapper`. A new sample called `Samples.Dapper` can be found in the `samples` folder in this repository.

@DataDog/apm-dotnet

### CODE REVIEW

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp [hunk 1]
```diff
@@ -31,9 +31,11 @@ CorProfiler::Initialize(IUnknown* cor_profiler_info_unknown) {
   const auto debug_enabled_value =
       GetEnvironmentValue(environment::debug_enabled);
 
-  if (debug_enabled_value == "1"_W || debug_enabled_value == "true"_W) {
-    debug_logging_enabled = true;
-  }
+  //if (debug_enabled_value == "1"_W || debug_enabled_value == "true"_W) {
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: Don't forget to remove this debugging code. Thanks!

File: samples/Samples.Npgsql/Program.cs [hunk 2]
```diff
@@ -26,45 +23,9 @@ private static async Task Main()
                     executeReaderWithBehaviorAsync: null
                 );
 
-
-                await testQueries.RunAsync();
-            }
-            */
-
-            using (var connection = CreateConnection())
```
[the remark this answers is not in the document]
Comment (bobuva) [reply samples/Samples.Npgsql/Program.cs]: ok, I thought it was just until the provider library was supported. Will do!

File: src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp (hunk 1 above)
Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp]: oops :)

File: samples/Samples.Dapper/Samples.Dapper.csproj [hunk 3]
```diff
@@ -0,0 +1,52 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>netcoreapp3.1</TargetFramework>
```
[the remark this answers is not in the document]
Comment (bobuva) [reply samples/Samples.Dapper/Samples.Dapper.csproj]: Sounds good - I'll make the change instead of just committing the suggestion.

File: samples/Samples.Dapper/Samples.Dapper.csproj [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  </PropertyGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net461'">
```
[the remark this answers is not in the document]
Comment (bobuva) [reply samples/Samples.Dapper/Samples.Dapper.csproj]: I'm not sure why Visual Studio decided to write the package references like that. It was automatic from installing via the Package Manager Console. I'll make the changes and retest.

[the remark this answers is not in the document]
Comment (bobuva) [reply samples/Samples.Dapper/Samples.Dapper.csproj]: Actually, I had installed the `1.60.6` version of `Dapper` per what a user had reported using. Apparently the Package Manager decided to update to use the later package.

File: src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj [hunk 5]
```diff
@@ -24,19 +24,19 @@ Users who need manual instrumentation should reference the "Datadog.Trace" packa
 
   <!-- For VS testing purposes only, copy all implementation assemblies to the
        to test the loading of managed profiler assemblies from disk -->
-  <PropertyGroup Condition="'$(BuildingInsideVisualStudio)'=='true' and '$(TargetFramework)' == 'netstandard2.0'">
+  <PropertyGroup Condition="'$(BuildingInsideVisualStudio)'=='true' and (( '$(TargetFramework)' == 'netstandard2.0' ) or (( '$(TargetFramework)' == 'netstandard2.1' ))) ">
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: I'm not sure if the `netstandard2.1` changes are necessary since we're only building for `netstandard2.0` at the moment

Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: Correct. Good catch, meant to remove that.

File: samples/Samples.Dapper/Program.cs [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+            {
+                var host = Environment.GetEnvironmentVariable("POSTGRES_HOST") ?? "localhost";
+                connectionString = $"Host={host};Username=postgres;Password=postgres;Database=postgres";
+            }
+
+            return new NpgsqlConnection(connectionString);
+        }
+    }
+
+    public class DapperTestHarness<TConnection, TCommand, TDataReader>
```
Comment (zacharycmontoya) [inline samples/Samples.Dapper/Program.cs]: What do you think about moving this `DapperTestHarness` class to the `Samples.DatabaseHelper` library, like the `RelationalDatabaseTestHarness` class? I'm thinking that we could run Dapper over all the other database clients as well to make sure we still trace those, though that work could be separated out into a different PR

Comment (bobuva) [reply samples/Samples.Dapper/Program.cs]: Yes, I think it is better there; meant to re-visit that!

File: src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj [hunk 7]
```diff
@@ -32,11 +32,11 @@ Users who need manual instrumentation should reference the "Datadog.Trace" packa
     <ProjectReference Include="..\Datadog.Trace\Datadog.Trace.csproj" />
   </ItemGroup>
 
-  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+  <ItemGroup Condition="' $(TargetFramework)' == 'netstandard2.0'">
```
Comment (zacharycmontoya) [inline src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: Not trying to be super nitpicky, but I'm worried that the introduced space here could make this equality false. Can you double check this?

File: samples/Samples.Dapper/Program.cs [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+using System.Data;
+using System.Threading.Tasks;
+using Npgsql;
+using Samples.DatabaseHelper;
+
+namespace Samples.Dapper
+{
+    internal static class Program
+    {
+        private static async Task Main()
```
Comment (zacharycmontoya) [inline samples/Samples.Dapper/Program.cs]: Super clean! 👍

File: src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj (hunk 7 above)
Comment (bobuva) [reply src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: Sure!

File: samples/Samples.Dapper/Samples.Dapper.csproj [hunk 9]
```diff
@@ -0,0 +1,22 @@
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+     <!-- override to remove net452 -->
+    <TargetFrameworks Condition="'$(OS)' == 'Windows_NT'">net461;netcoreapp2.1;netcoreapp3.0;netcoreapp3.1</TargetFrameworks>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Dapper" Version="2.0.30" />
```
Comment (zacharycmontoya) [inline samples/Samples.Dapper/Samples.Dapper.csproj]: nit: I think we can remove this Dapper reference now that it's encapsulated in the helper library

File: src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj [hunk 10]
```diff
@@ -32,11 +32,11 @@ Users who need manual instrumentation should reference the "Datadog.Trace" packa
     <ProjectReference Include="..\Datadog.Trace\Datadog.Trace.csproj" />
   </ItemGroup>
 
-  <ItemGroup Condition=" '$(TargetFramework)' == 'netstandard2.0' ">
+  <ItemGroup Condition="'$(TargetFramework)' == 'netstandard2.0'">
     <PackageReference Include="System.Net.Http" Version="4.3.4" />
   </ItemGroup>
 
-  <ItemGroup Condition=" '$(TargetFramework)' != 'netstandard2.0' ">
+  <ItemGroup Condition=" '$(TargetFramework)' != 'netstandard2.0'">
```
Comment (lucaspimentel) [inline src/Datadog.Trace.ClrProfiler.Managed/Datadog.Trace.ClrProfiler.Managed.csproj]: Missed one 😉 
```suggestion
  <ItemGroup Condition="'$(TargetFramework)' != 'netstandard2.0'">
```
